enterprise/coderd/userauth_test.go

```diff
@@ -352,6 +355,213 @@ func TestUserOIDC(t *testing.T) {
+			},
+			initialOrgGroups:   []string{},
+			expectedUserGroups: []string{},
+			expectedOrgGroups:  []string{},
+			claims:             jwt.MapClaims{},
+		},
+		{
+			name: "GroupSyncDisabled",
+			modCfg: func(cfg *coderd.OIDCConfig) {
+				// Disable group sync
+				cfg.GroupField = ""
+				cfg.GroupFilter = regexp.MustCompile(".*")
+			},
+			initialOrgGroups:   []string{"a", "b", "c", "d"},
+			initialUserGroups:  []string{"b", "c", "d"},
+			expectedUserGroups: []string{"b", "c", "d"},
+			expectedOrgGroups:  []string{"a", "b", "c", "d"},
+			claims:             jwt.MapClaims{},
+		},
+		{
+			// From a,c,b -> b,c,d
+			name: "ChangeUserGroups",
+			modCfg: func(cfg *coderd.OIDCConfig) {
+				cfg.GroupMapping = map[string]string{
+					"D": "d",
+				}
+			},
+			initialOrgGroups:   []string{"a", "b", "c", "d"},
+			initialUserGroups:  []string{"a", "b", "c"},
+			expectedUserGroups: []string{"b", "c", "d"},
+			expectedOrgGroups:  []string{"a", "b", "c", "d"},
+			claims: jwt.MapClaims{
+				// D -> d mapped
+				"groups": []string{"b", "c", "D"},
+			},
+		},
+		{
+			// From a,c,b -> []
+			name: "RemoveAllGroups",
+			modCfg: func(cfg *coderd.OIDCConfig) {
+				cfg.GroupFilter = regexp.MustCompile(".*")
+			},
+			initialOrgGroups:   []string{"a", "b", "c", "d"},
+			initialUserGroups:  []string{"a", "b", "c"},
+			expectedUserGroups: []string{},
+			expectedOrgGroups:  []string{"a", "b", "c", "d"},
+			claims:             jwt.MapClaims{
+				// No claim == no groups
+			},
+		},
+		{
+			// From a,c,b -> b,c,d,e,f
+			name: "CreateMissingGroups",
+			modCfg: func(cfg *coderd.OIDCConfig) {
+				cfg.CreateMissingGroups = true
+			},
+			initialOrgGroups:   []string{"a", "b", "c", "d"},
+			initialUserGroups:  []string{"a", "b", "c"},
+			expectedUserGroups: []string{"b", "c", "d", "e", "f"},
+			expectedOrgGroups:  []string{"a", "b", "c", "d", "e", "f"},
+			claims: jwt.MapClaims{
+				"groups": []string{"b", "c", "d", "e", "f"},
+			},
+		},
+		{
+			// From a,c,b -> b,c,d,e,f
+			name: "CreateMissingGroupsFilter",
+			modCfg: func(cfg *coderd.OIDCConfig) {
+				cfg.CreateMissingGroups = true
+				// Only single letter groups
+				cfg.GroupFilter = regexp.MustCompile("^[a-z]$")
+				cfg.GroupMapping = map[string]string{
+					// Does not match the filter, but does after being mapped!
+					"zebra": "z",
+				}
+			},
+			initialOrgGroups:   []string{"a", "b", "c", "d"},
+			initialUserGroups:  []string{"a", "b", "c"},
+			expectedUserGroups: []string{"b", "c", "d", "e", "f", "z"},
+			expectedOrgGroups:  []string{"a", "b", "c", "d", "e", "f", "z"},
+			claims: jwt.MapClaims{
+				"groups": []string{
+					"b", "c", "d", "e", "f",
+					// These groups are ignored
+					"excess", "ignore", "dumb", "foobar", "zebra",
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := testutil.Context(t, testutil.WaitLong*40000)
```

Comment:
This seems a bit long

Reply:
Whoooops. I do this because when I use a debugger the contexts always expire lol.